docking/dock_and_equilibrate.py

```diff
@@ -0,0 +1,387 @@
+        for step, lamb in enumerate(insertion_lambda_schedule):
+            ctxt.step(lamb)
+            if step % 100 == 0:
+                report_step(ctxt, step, lamb, combined_bps, host_box, u_impls, guest_name, insertion_steps, "INSERTION")
+                if not fewer_outfiles:
+                    host_coords = ctxt.get_x_t()[: len(solvated_host_coords)] * 10
+                    guest_coords = ctxt.get_x_t()[len(solvated_host_coords) :] * 10
+                    write_frame(
+                        host_coords,
+                        solvated_host_mol,
+                        guest_coords,
+                        guest_mol,
+                        outdir,
+                        guest_name,
+                        str(step).zfill(len(str(insertion_steps))),
+                        f"ins",
+                    )
+                if too_much_force(ctxt, combined_bps, host_box, u_impls, lamb):
+                    calc_work = False
+                    break
+
+        if (
+            abs(du_dl_obs.full_du_dl()[0]) > 0.001
+            or abs(du_dl_obs.full_du_dl()[-1]) > 0.001
+        ):
+            print("Error: du_dl endpoints are not ~0")
+            calc_work = False
+
+        if calc_work:
+            work = np.trapz(
+                du_dl_obs.full_du_dl(), insertion_lambda_schedule[::subsample_freq]
+            )
+            print(f"guest_name: {guest_name}\tinsertion_work: {work:.2f}")
+
+        # equilibrate
+        for step in range(eq_steps):
+            ctxt.step(0.00)
+            if step % 1000 == 0:
+                report_step(ctxt, step, 0.00, combined_bps, host_box, u_impls, guest_name, eq_steps, 'EQUILIBRATION')
+                host_coords = ctxt.get_x_t()[: len(solvated_host_coords)] * 10
+                guest_coords = ctxt.get_x_t()[len(solvated_host_coords) :] * 10
+                write_frame(
+                    host_coords,
+                    solvated_host_mol,
+                    guest_coords,
+                    guest_mol,
+                    outdir,
+                    guest_name,
+                    str(step).zfill(len(str(eq_steps))),
+                    f"eq",
+                )
+                if too_much_force(ctxt, combined_bps, host_box, u_impls, 0.0):
+                    break
+
+        end_time = time.time()
+        print(f"{guest_name} took {(end_time - start_time):.2f} seconds")
+
+
+def report_step(ctxt, step, lamb, bps, box, u_impls, guest_name, n_steps, stage):
+    l_energies = []
+    names = []
+    for name, impl in zip(bps, u_impls):
+        _, _, u = impl.execute(ctxt.get_x_t(), box, lamb)
+        l_energies.append(u)
+        names.append(name)
+        energy = sum(l_energies)
+
+    print(
+        f"{stage}\t"
+        f"guest_name: {guest_name}\t"
+        f"step: {str(step).zfill(len(str(n_steps)))}\t"
+        f"lambda: {lamb:.2f}\t"
+        f"energy: {energy:.2f}"
+    )
+
+
+def too_much_force(ctxt, bps, box, u_impls, lamb):
```

Comment:
This function should be called sparingly. It is *hideously* expensive due to all the memcpys. Even once every 100 steps might be too often.